README.md
XiVO confd
==========

[![Build Status](https://jenkins.wazo.community/buildStatus/icon?job=xivo-confd)](https://jenkins.wazo.community/job/xivo-confd)

XiVO CONFD is a web server that provides a [RESTful](http://en.wikipedia.org/wiki/Representational_state_transfer)
service for configuring and managing a Wazo server. Further details on how to use the API can be found on
the [Wazo API web site](http://api.wazo.community)


Installing xivo-confd
---------------------

The server is already provided as a part of [Wazo](http://documentation.wazo.community).
Please refer to [the documentation](http://documentation.wazo.community/en/stable/installation/installsystem.html) for
further details on installing one.


Running unit tests
------------------

```
apt-get install libpq-dev python-dev libffi-dev libyaml-dev
pip install tox
tox --recreate -e py27
```

Running integration tests
-------------------------

You need Docker installed.

```
cd integration_tests
pip install -U -r test-requirements.txt
make test-setup
make test
```


Development
-----------

### Modified database

You need the repos xivo-manage-db up-to-date.

1. ```git clone https://github.com/wazo-pbx/xivo-manage-db```
2. ```MANAGE_DB_DIR=../../xivo-manage-db make update-db```
3. Execute the steps above to run integration tests


### Modified xivo-provisioning

You need the repos xivo-provisioning up-to-date.

1. ```git clone https://github.com/wazo-pbx/xivo-provisioning```
2. ```PROVD_DIR=../../xivo-provisioning make build-provd```
3. Execute the steps above to run integration tests


### Mounting libraries

In case you need to mount libraries (xivo-dao, xivo-bus, lib-python) inside the xivo-confd container:

1. Uncomment the confd volumes in ```integration_tests/assets/base/docker-compose.yml```
2. Set the environment variable: ```export LOCAL_GIT_REPOS=/parent/directory/to/all/git/repos```
3. Execute the steps above to run integration tests


Debuging SQL queries
====================

* Add custom configuration:
```
rest_api:
  debug_db: True
```
* Check the log


Profiling
=========

Terminal
--------

* add custom configuration:
```
rest_api:
  profile: True
```
* Check the log


Graphical
---------

* ```pip install gprof2dot```
* ```apt-get install graphviz```
* add custom configuration:
```
rest_api:
  profile: '/tmp/confd-profiling'
```
* process file in directory with the following command:

```gprof2dot -f pstats <directory>/<file> | dot -Tpng -o output.png```


CHANGELOG.md
Changelog
=========

19.15
-----

* A new read-only parameter has been added to the users resource:

  * `call_pickup_targets`


19.13
-----

* A new API for configuring features access has been added:

  * GET `/1.1/access_features`
  * POST `/1.1/access_features`
  * GET `/1.1/access_features/<access_feature_id>`
  * PUT `/1.1/access_features/<access_features_id>`
  * DELETE `/1.1/access_features/<access_feature_id>`


19.12
-----

* The API `/1.1/call_logs` for call logs have been removed


19.11
-----

* A new API for configuring the provisioning registrars has been added:

  * GET `/1.1/registrars`
  * POST `/1.1/registrars`
  * GET `/1.1/registrars/<registrar_id>`
  * PUT `/1.1/registrars/<registrar_id>`
  * DELETE `/1.1/registrars/<registrar_id>`


19.10
-----

* A new API for associating a line with an application has been added:

  * PUT `/1.1/lines/<line_id>/applications/<application_id>`
  * DELETE `/1.1/lines/<line_id>/applications/<application_id>`

* For newly created applications with `node` destination, calls are not
  automatically answered anymore. A new `answer` parameter has been
  added in `destination_options` to configure the answer behavior of the node.


19.08
-----

* Due to the removal of the Wazo Client QT, the following endpoints has been deleted:
  * GET `/1.1/cti_profiles`
  * GET `/1.1/cti_profiles/<cti_profile_id>`
  * GET `/1.1/users/<user_uuid>/cti`
  * PUT `/1.1/users/<user_uuid>/cti`
  Also note that endpoints `/users` don't have `cti_profile` parameter anymore.


19.07
-----

* The following endpoints are now multi-tenant.

  This means that created resources will be in the same tenant as the creator or in the tenant
  specified by the Wazo-Tenant HTTP header. Listing resources will also only list the ones in the
  user's tenant unless a sub-tenant is specified using the Wazo-Tenant header. The `recurse=true`
  query string can be used to list from multiple tenants. GET, DELETE and PUT on a resource that is
  not tenant accessible will result in a 404. New readonly parameter has also been added:
  `tenant_uuid`.

  * `/1.1/queues/skillrules`

* Added Unallocated devices endpoints:
  * GET `/1.1/devices/unallocated`
  * PUT `/1.1/devices/unallocated/<device_id>`

* The following endpoints are now multi-tenant.

  This means that created resources will be in the same tenant as the creator or in the tenant
  specified by the Wazo-Tenant HTTP header. Listing resources will also only list the ones in the
  user's tenant unless a sub-tenant is specified using the Wazo-Tenant header. The `recurse=true`
  query string can be used to list from multiple tenants. GET, DELETE and PUT on a resource that is
  not tenant accessible will result in a 404. New readonly parameter has also been added:
  `tenant_uuid`.

  * `/1.1/agents/skills`


19.06
-----

* The `http` configuration section has been removed. Also the `https` section has moved to
  `rest_api` to be similar with other daemons.
* The `/1.1/users/import` PUT has been disabled since it creates invalid configurations
* The `/1.1/users/export` resource will now only list users from the specified tenant.

* The following endpoints are now multi-tenant.

  This means that created resources will be in the same tenant as the creator or in the tenant
  specified by the Wazo-Tenant HTTP header. Listing resources will also only list the ones in the
  user's tenant unless a sub-tenant is specified using the Wazo-Tenant header. The `recurse=true`
  query string can be used to list from multiple tenants. GET, DELETE and PUT on a resource that is
  not tenant accessible will result in a 404. New readonly parameter has also been added:
  `tenant_uuid`.

  * `/1.1/agents`


19.05
-----

* The following endpoints are now multi-tenant.

  This means that created resources will be in the same tenant as the creator or in the tenant
  specified by the Wazo-Tenant HTTP header. Listing resources will also only list the ones in the
  user's tenant unless a sub-tenant is specified using the Wazo-Tenant header. The `recurse=true`
  query string can be used to list from multiple tenants. GET, DELETE and PUT on a resource that is
  not tenant accessible will result in a 404. New readonly parameter has also been added:
  `tenant_uuid`.

  * `/1.1/queues`
  * `/1.1/switchboards`

* Added Provisioning Networking Configuration endpoints:

  * GET `/1.1/provisioning/networking`
  * PUT `/1.1/provisioning/networking`

* New readonly parameters have been added to the device resource:

  * `is_new`

* The `cti_profile_name`, `cti_profile_enabled` and `entity` fields have been removed from the `/1.1/users/import` resource.
* The `/1.1/users/import` will now create the users in the creator's tenant or the specified tenant.
* The `/1.1/user/<user_id>/entities` resource has been removed.
* The `/1.1/entities` resource has been removed.
* The `summary` view does not return the entity anymore.
* The following fields have been removed from the `/1.1/wizard` POST body

  * `entity_name`
  * `context_outcall`
  * `context_incall`
  * `context_internal`

* The following steps have been removed from the `/1.1/wizard` POST body

  * `phonebook`
  * `tenant`

* In `/dhcp`, the field `extra_network_interfaces` is renamed to `network_interfaces`.
* A new function key type has been added: `groupmember`


19.04
-----

* New readonly parameters have been added to the funckeys resource:

  * For destinations of type `bsfilter`:

      * `filter_member_firstname`
      * `filter_member_lastname`

* New readonly parameters have been added to the call filter surrogate resource:

  * `member_id`

* The following endpoints are now multi-tenant.

  This means that created resources will be in the same tenant as the creator or in the tenant
  specified by the Wazo-Tenant HTTP header. Listing resources will also only list the ones in the
  user's tenant unless a sub-tenant is specified using the Wazo-Tenant header. The `recurse=true`

xivo_confd/http_server.py
# -*- coding: utf-8 -*-
# Copyright 2016-2018 The Wazo Authors  (see the AUTHORS file)
# SPDX-License-Identifier: GPL-3.0+

import os
import urllib
import logging
import cherrypy

from cherrypy.process.servers import ServerAdapter
from cheroot import wsgi
from flask import Flask, g, request
from flask_cors import CORS
from flask_restful import Api
from sqlalchemy.exc import SQLAlchemyError
from werkzeug.contrib.profiler import ProfilerMiddleware
from werkzeug.contrib.fixers import ProxyFix

from xivo import http_helpers
from xivo.http_helpers import ReverseProxied
from xivo_dao.helpers.db_manager import Session
from xivo_dao.helpers.db_utils import session_scope
from xivo_dao.resources.infos import dao as info_dao

from ._bus import BusPublisher
from ._sysconfd import SysconfdPublisher
from .debug import EngineDebuggingSignalEvents, get_debug_queries
from .helpers.converter import FilenameConverter

logger = logging.getLogger(__name__)
app = Flask('xivo_confd')
api = Api(app, prefix="/1.1")
_do_not_log_data_endpoints = []
cherrypy.engine.signal_handler.set_handler('SIGTERM', cherrypy.engine.exit)


def add_endpoint_to_do_not_log_data_list(endpoint):
    # XXX name is bad
    _do_not_log_data_endpoints.append(endpoint)


def get_bus_publisher():
    publisher = g.get('bus_publisher')
    if not publisher:
        publisher = g.bus_publisher = BusPublisher.from_config(app.config)
    return publisher


def get_sysconfd_publisher():
    publisher = g.get('sysconfd_publisher')
    if not publisher:
        publisher = g.sysconfd_publisher = SysconfdPublisher.from_config(app.config)
    return publisher


def log_requests():
    url = request.url.encode('utf8')
    url = urllib.unquote(url)
    params = {
        'method': request.method,
        'url': url,
    }
    if request.data and request.endpoint not in _do_not_log_data_endpoints:
        params.update({'data': request.data})
        logger.info("%(method)s %(url)s with data %(data)s ", params)
    else:
        logger.info("%(method)s %(url)s", params)


def after_request(response):
    commit_database()
    flush_sysconfd()
    flush_bus()
    return http_helpers.log_request(response)


def commit_database():
    try:
        Session.commit()
    except SQLAlchemyError:
        Session.rollback()
        raise
    finally:
        Session.remove()


def log_database_queries(response):
    for query in get_debug_queries():
        logger.debug("\nQUERY: %s\nParameters: %s\nDuration: %fs\n"
                     % (query.statement, query.parameters, query.duration))
    return response


def flush_sysconfd():
    publisher = g.get('sysconfd_publisher')
    if publisher:
        publisher.flush()


def flush_bus():
    publisher = g.get('bus_publisher')
    if publisher:
        publisher.flush()


def load_uuid():
    with session_scope():
        app.config['uuid'] = info_dao.get().uuid


class HTTPServer(object):

    def __init__(self, global_config):
        self.config = global_config['rest_api']
        http_helpers.add_logger(app, logger)

        app.before_first_request(load_uuid)
        app.before_request(log_requests)
        app.after_request(after_request)

        app.secret_key = os.urandom(24)
        app.url_map.converters['filename'] = FilenameConverter

        app.config.update(global_config)
        app.config['MAX_CONTENT_LENGTH'] = 40 * 1024 * 1024

        self._load_cors()

    def _load_cors(self):
        cors_config = dict(self.config.get('cors', {}))
        enabled = cors_config.pop('enabled', False)
        if enabled:
            CORS(app, **cors_config)

    def run(self):
        http_config = self.config['http']
        https_config = self.config['https']

        cherrypy.engine.signal_handler.set_handler('SIGTERM', cherrypy.engine.exit)

        if self.config['debug_db']:
            EngineDebuggingSignalEvents(Session.get_bind(), app.import_name).register()
            app.after_request(log_database_queries)

        if self.config['profile']:
            from xivo.xivo_logging import _StreamToLogger
            stream = _StreamToLogger(logger, logging.DEBUG)
            profile_dir = None
            restrictions = [30]  # Display only the 30 slower methods
            if isinstance(self.config['profile'], unicode):
                profile_dir = self.config['profile']
                restrictions = []
            app.wsgi_app = ProfilerMiddleware(
                app.wsgi_app,
                profile_dir=profile_dir,
                restrictions=restrictions,
                stream=stream
            )

        wsgi_app = ReverseProxied(ProxyFix(wsgi.WSGIPathInfoDispatcher({'/': app})))

        cherrypy.server.unsubscribe()
        cherrypy.config.update({'environment': 'production'})

        if not (http_config['enabled'] or https_config['enabled']):
            logger.critical('No HTTP/HTTPS server enabled')
            return

        if https_config['enabled']:
            try:
                bind_addr_https = (https_config['listen'], https_config['port'])
                server_https = wsgi.WSGIServer(bind_addr=bind_addr_https,
                                               wsgi_app=wsgi_app)
                server_https.ssl_adapter = http_helpers.ssl_adapter(https_config['certificate'],
                                                                    https_config['private_key'])
                ServerAdapter(cherrypy.engine, server_https).subscribe()

                logger.debug('HTTPS server starting on %s:%s', *bind_addr_https)

            except IOError as e:
                logger.warning("HTTPS server won't start: %s", e)
        else:
            logger.debug('HTTPS server is disabled')

        if http_config['enabled']:
            bind_addr_http = (http_config['listen'], http_config['port'])
            server_http = wsgi.WSGIServer(bind_addr=bind_addr_http,
                                          wsgi_app=wsgi_app)
            ServerAdapter(cherrypy.engine, server_http).subscribe()

            logger.debug('HTTP server starting on %s:%s', *bind_addr_http)
        else:
            logger.debug('HTTP server is disabled')

        cherrypy.engine.start()
        cherrypy.engine.block()


xivo_confd/config.py
# -*- coding: UTF-8 -*-
# Copyright 2013-2018 The Wazo Authors  (see the AUTHORS file)
# SPDX-License-Identifier: GPL-3.0+

import argparse
import yaml

from xivo.chain_map import ChainMap
from xivo.config_helper import parse_config_file, read_config_file_hierarchy
from xivo.xivo_logging import get_log_level_by_name

API_VERSION = '1.1'
WIZARD_KEY_FILE = '/var/lib/xivo-auth-keys/xivo-wizard-key.yml'

DEFAULT_CONFIG = {
    'debug': False,
    'user': 'www-data',
    'log_level': 'info',
    'config_file': '/etc/xivo-confd/config.yml',
    'extra_config_files': '/etc/xivo-confd/conf.d/',
    'log_filename': '/var/log/xivo-confd.log',
    'pid_filename': '/var/run/xivo-confd/xivo-confd.pid',
    'rest_api': {
        'profile': False,
        'debug_db': False,
        'http': {
            'enabled': True,
            'listen': '127.0.0.1',
            'port': 9487,
        },
        'https': {
            'enabled': True,
            'listen': '0.0.0.0',
            'port': 9486,
            'certificate': '/usr/share/xivo-certs/server.crt',
            'private_key': '/usr/share/xivo-certs/server.key',
        },
        'cors': {
            'enabled': True,
            'allow_headers': ['Content-Type'],
        },
    },
    'auth': {
        'host': 'localhost',
        'port': 9497,
        'verify_certificate': '/usr/share/xivo-certs/server.crt',
        'key_file': '/var/lib/xivo-auth-keys/xivo-confd-key.yml',
    },
    'ari': {
        'host': 'localhost',
        'port': 5039,
        'https': False,
        'username': 'xivo',
        'password': 'Nasheow8Eag',
    },
    'bus': {
        'username': 'guest',
        'password': 'guest',
        'host': 'localhost',
        'port': 5672,
        'exchange_name': 'xivo',
        'exchange_type': 'topic',
        'exchange_durable': True,
    },
    'consul': {
        'scheme': 'https',
        'host': 'localhost',
        'port': 8500,
        'verify': '/usr/share/xivo-certs/server.crt',
    },
    'dird': {
        'host': 'localhost',
        'port': 9489,
        'verify_certificate': '/usr/share/xivo-certs/server.crt',
    },
    'provd': {
    },
    'sysconfd': {
        'host': 'localhost',
        'port': '8668'
    },
    'enabled_plugins': {
        'api': True,
        'call_log': True,
        'call_filter': True,
        'call_filter_fallback': True,
        'call_filter_user': True,
        'call_permission': True,
        'confbridge': True,
        'conference': True,
        'conference_extension': True,
        'configuration': True,
        'context': True,
        'cti_profile': True,
        'device': True,
        'endpoint_custom': True,
        'endpoint_iax': True,
        'endpoint_sccp': True,
        'endpoint_sip': True,
        'entity': True,
        'extension': True,
        'extension_feature': True,
        'features': True,
        'func_key': True,
        'group': True,
        'group_call_permission': True,
        'group_extension': True,
        'group_fallback': True,
        'group_member_user': True,
        'group_schedule': True,
        'iax_callnumberlimits': True,
        'iax_general': True,
        'incall': True,
        'incall_extension': True,
        'incall_schedule': True,
        'info': True,
        'ivr': True,
        'line': True,
        'line_device': True,
        'line_endpoint': True,
        'line_extension': True,
        'line_sip': True,
        'moh': True,
        'outcall': True,
        'outcall_call_permission': True,
        'outcall_extension': True,
        'outcall_schedule': True,
        'outcall_trunk': True,
        'paging': True,
        'paging_user': True,
        'parking_lot': True,
        'parking_lot_extension': True,
        'queue_member': True,
        'register_iax': True,
        'register_sip': True,
        'schedule': True,
        'sccp_general': True,
        'sip_general': True,
        'sound': True,
        'sound_language': True,
        'switchboard': True,
        'switchboard_member': True,
        'timezone': True,
        'trunk': True,
        'trunk_endpoint': True,
        'trunk_register': True,
        'user': True,
        'user_agent': True,
        'user_call_permission': True,
        'user_cti_profile': True,
        'user_entity': True,
        'user_fallback': True,
        'user_group': True,
        'user_import': True,
        'user_line': True,
        'user_line_associated': True,
        'user_schedule': True,
        'user_voicemail': True,
        'voicemail': True,
        'voicemail_general': True,
        'voicemail_zonemessages': True,
        'wizard': True,
    },
    'service_discovery': {
        'enabled': True,
        'advertise_address': 'localhost',
        'advertise_port': 9486,
        'advertise_address_interface': 'eth0',
        'refresh_interval': 25,
        'retry_interval': 2,
        'ttl_interval': 30,
        'extra_tags': [],
    },
    'wizard': {
        'service_id': None,
        'service_key': None,
    },
}


def load(argv):
    try:
        with open(WIZARD_KEY_FILE, 'r') as f:
            key_config = {'wizard': yaml.load(f)}
    except IOError:
        key_config = {}

    cli_config = _parse_cli_args(argv)
    file_config = read_config_file_hierarchy(ChainMap(cli_config, DEFAULT_CONFIG))
    reinterpreted_config = _get_reinterpreted_raw_values(ChainMap(cli_config, file_config, DEFAULT_CONFIG))
    service_key = _load_key_file(ChainMap(cli_config, file_config, DEFAULT_CONFIG))
    return ChainMap(reinterpreted_config, key_config, cli_config, service_key, file_config, DEFAULT_CONFIG)


def _load_key_file(config):
    key_file = parse_config_file(config['auth']['key_file'])
    return {'auth': {'username': key_file.get('service_id'),
                     'password': key_file.get('service_key')}}


def _parse_cli_args(argv):
    parser = argparse.ArgumentParser()
    parser.add_argument('-c',
                        '--config-file',
                        action='store',
                        help="The path where is the config file. Default: %(default)s")
    parser.add_argument('-d',
                        '--debug',
                        action='store_true',
                        help="Log debug messages. Overrides log_level. Default: %(default)s")
    parser.add_argument('-l',
                        '--log-level',
                        action='store',
                        default='INFO',
                        help="Logs messages with LOG_LEVEL details. Must be one of:\n"
                             "critical, error, warning, info, debug. Default: %(default)s")
    parser.add_argument('-u',
                        '--user',
                        action='store',
                        help="The owner of the process.")
    parser.add_argument('-p',
                        '--profile',
                        help="Write profiling stats to directory (for debugging performance issues)",
                        action='store')
    parsed_args = parser.parse_args(argv)

    result = {}
    if parsed_args.config_file:
        result['config_file'] = parsed_args.config_file
    if parsed_args.debug:
        result['debug'] = parsed_args.debug
    if parsed_args.profile:
        result['rest_api']['profile'] = parsed_args.profile
    if parsed_args.log_level:
        result['log_level'] = parsed_args.log_level
    if parsed_args.user:
        result['user'] = parsed_args.user

    return result


def _get_reinterpreted_raw_values(config):
    result = {}

    log_level = config.get('log_level')
    if log_level:
        result['log_level'] = get_log_level_by_name(log_level)

    return result
